misc/codegen/generators/qlgen.py

```diff
@@ -312,7 +312,43 @@ def _get_stub(cls: schema.Class, base_import: str, generated_import_prefix: str)
     else:
         accessors = []
     return ql.Stub(name=cls.name, base_import=base_import, import_prefix=generated_import_prefix,
-                   synth_accessors=accessors, ql_internal="ql_internal" in cls.pragmas)
+                   doc=cls.doc, synth_accessors=accessors,
+                   ql_internal="ql_internal" in cls.pragmas)
+
+
+def _patch_class_qldocs(cls: str, qldoc: str, stub_file: pathlib.Path):
+    if not qldoc or not stub_file.exists():
+        return
+    qldoc = "\n".join(l.rstrip() for l in qldoc.splitlines())
+    tmp = stub_file.with_suffix(f'{stub_file.suffix}.bkp')
+    header = "// the following QLdoc is generated: if you need to edit it, do it in the schema file\n"
+    with open(stub_file) as input:
+        qldoc_start = None
+        qldoc_end = None
+        class_start = None
+        for lineno, line in enumerate(input, 1):
+            if line == header:
+                qldoc_start = lineno
+            if line.startswith("/**") and lineno - 1 != qldoc_start:
+                qldoc_start = lineno
+            if line.endswith(" */\n"):
+                qldoc_end = lineno + 1
+            elif line.startswith(f"class {cls}"):
+                class_start = lineno
+                break
```

Comment:
I'm a bit worried about edge cases, where one or more of these cases fails to match the line we expect it to for some reason, and assumptions are broken thereafter break.  For example (everything `<--` and after is annotations, not part of the example):
```
/**<-- qldoc_start
 * A class.
 */<-- qldoc_end
class foo { }

/** <-- there's an extra space crept in on the end of this line, so it's not a qldoc_start
 * Another class.
 */<-- qldoc_end
class bar { } <-- class_start
```
In this case we may delete the entire `class foo` along with the two qldocs and replace it with the comment for `class bar`.  Except that this example won't get through the autoformatter - but maybe something else can?

Reply:
Nested classes could be another problematic area.  For example:
```
/**
 * A class.
 */
class foo {
  class bar {
  }
}
``` 
Though I guess we currently wouldn't recognize the indented class as a `class_start` anyway.

Reply:
QL doesn't have a concept of "nested classes" so this is probably not a problem.

Reply:
about
```codeql
/** <-- there's an extra space crept in on the end of this line, so it's not a qldoc_start
```
the test is `line.startswith("/**")`, so that would be covered. In fact, the current code also works in one line QLdocs like
```codeql
/** bla bla */
class foo ...
```
(though it will always overwrite the qldoc with one across multiple lines)

But anyway, there might be an assumption, which is that the code is formatted: this might be a weak point, so maybe we should autoformat the stubs before trying to patch them. For example, your example would indeed clear things if the space crept _before_ the docstring (which is a formatting violation):
```codeql
 /**<-- there's an extra space crept in on at the beginning of this line, so it's not a qldoc_start
  * Another class.
  */<-- qldoc_end
class bar { } <-- class_start
```

 Another safety measure might be to clear `qldoc_end` when we meet a `qldoc_start` and assert `qldoc_end` is cleared before we try to set it.

Reply:
hmm, autoformatting would probably be a bit too slow, but we can put in more failsafes. For example, we could check that all lines between a `qldoc_start` and `qldoc_end` start with ` *`. We would still error out on some unformatted stubs, but at least we would not be deleting proper code.

Reply:
I think putting in an extra failsafe or two, and a comment explaining the assumptions here would be make me happy.  The assumptions in particular are that we only need to replace one comment in the file, that it's on a class that is not indented (i.e. not in a `module`), and that the file is (at least approximately) in the style of our autoformatter.

> QL doesn't have a concept of "nested classes" so this is probably not a problem.

Yeah, I forgot the target is QL code not Swift code.  Thanks for pointing this out.

Reply:
I've opted for a multiline regex with non-greedy `.*?` to do the dirty work, which is much more concise, understandable and robust than the complex code I originally came up with.

Reply:
Yes I think the new approach is much more robust (assuming the regex works correctly in practice -  it *looks* good to me).  Thank you.